x/smart/stream_dialer.go

```diff
@@ -250,21 +250,31 @@ func (f *StrategyFinder) testDialer(ctx context.Context, dialer transport.Stream
 		testAddr := net.JoinHostPort(testDomain, "443")
 		f.logCtx(ctx, "🏃 running test: '%v' (domain: %v)\n", transportCfg, testDomain)
 
-		ctx, cancel := context.WithTimeout(ctx, f.TestTimeout)
+		testCtx, cancel := context.WithTimeout(ctx, f.TestTimeout)
 		defer cancel()
-		testConn, err := dialer.DialStream(ctx, testAddr)
+
+		timeoutChan := make(chan struct{})
+		go func() {
+			<-testCtx.Done()
+			if errors.Is(testCtx.Err(), context.DeadlineExceeded) {
+				f.logCtx(testCtx, "⏱️ dialer failure, test timed out: '%v' (domain: %v), duration=%v ❌\n", transportCfg, testDomain, time.Since(startTime))
+				close(timeoutChan)
+			}
+		}()
+
+		testConn, err := dialer.DialStream(testCtx, testAddr)
 		if err != nil {
-			f.logCtx(ctx, "🏁 failed to dial: '%v' (domain: %v), duration=%v, dial_error=%v ❌\n", transportCfg, testDomain, time.Since(startTime), err)
```

Comment:
This is already logging dial errors, which should include timeouts? Why do you need the new code?

Reply:
no we don't, currently there's no logging when these tests timeout. We return an error, but any errors returned inside raceTests get eaten up, then raceTests just returns the error "all tests failed".

But maybe instead of this change the right fix is to add more logging when errors are returned [here](https://github.com/Jigsaw-Code/outline-sdk/blob/66412bf6501878c144e9d6680cd273ca52f6b2b5/x/smart/stream_dialer.go#L333) and [here](https://github.com/Jigsaw-Code/outline-sdk/blob/66412bf6501878c144e9d6680cd273ca52f6b2b5/x/smart/stream_dialer.go#L338) instead.

But we want to retain the existing behavior, which is that when one test succeeds we stop logging for the others.

Reply:
Pass the original testDialer context to this log call. That way it can still log if the race is not done, but the test is.